File: backend/src/handlers/requirements.rs
use axum::{
    extract::{Path, State},
    http::StatusCode,
    response::IntoResponse,
    Json,
};
use sqlx::PgPool;
use uuid::Uuid;
use validator::Validate;

use crate::dto::requirement_dto::{
    CreateConnectionRequest, CreateRequirementRequest, CreateStepRequest, UpdateRequirementRequest,
};
use crate::errors::AppError;
use crate::models::requirement::RequirementWithSteps;
use crate::repositories::requirement_repo;
use crate::utils::gherkin_parser::GherkinParser;

// Requirement handlers
pub async fn create_requirement(
    State(pool): State<PgPool>,
    Path(project_id): Path<Uuid>,
    Json(payload): Json<CreateRequirementRequest>,
) -> Result<impl IntoResponse, AppError> {
    tracing::info!(
        project_id = %project_id,
        title = %payload.title,
        "Creating new requirement"
    );
    tracing::debug!(
        payload = ?payload,
        "Requirement payload received"
    );

    payload
        .validate()
        .map_err(|e| {
            tracing::error!(
                error = %e,
                payload = ?payload,
                "Validation failed for requirement"
            );
            AppError::Validation(e.to_string())
        })?;

    tracing::debug!("Validating Gherkin syntax");
    // Validate Gherkin syntax
    GherkinParser::validate(&payload.gherkin_scenario)
        .map_err(|e| {
            tracing::error!(
                error = %e,
                gherkin = %payload.gherkin_scenario,
                "Gherkin validation failed"
            );
            AppError::Validation(format!("Invalid Gherkin: {}", e))
        })?;

    tracing::debug!("Creating requirement in database");
    // Create requirement
    let requirement = requirement_repo::create_requirement(
        &pool,
        project_id,
        &payload.title,
        payload.description.as_deref(),
        &payload.gherkin_scenario,
        payload.position_x,
        payload.position_y,
    )
    .await
    .map_err(|e| {
        tracing::error!(
            error = ?e,
            project_id = %project_id,
            title = %payload.title,
            "Failed to create requirement in database"
        );
        e
    })?;

    tracing::info!(
        requirement_id = %requirement.id,
        "Requirement created successfully"
    );

    // Parse Gherkin and create steps
    tracing::debug!("Parsing Gherkin scenario into steps");
    let parsed_steps = GherkinParser::parse(&payload.gherkin_scenario);
    tracing::debug!(
        step_count = parsed_steps.len(),
        "Parsed {} steps from Gherkin",
        parsed_steps.len()
    );

    for (idx, step) in parsed_steps.iter().enumerate() {
        tracing::debug!(
            step_index = idx,
            step_type = %step.step_type,
            step_text = %step.step_text,
            "Creating step"
        );

        let domain_refs = if step.domain_references.is_empty() {
            None
        } else {
            Some(serde_json::to_value(&step.domain_references).unwrap())
        };

        requirement_repo::create_step(
            &pool,
            requirement.id,
            &step.step_type,
            step.step_order,
            &step.step_text,
            domain_refs.as_ref(),
        )
        .await
        .map_err(|e| {
            tracing::error!(
                error = ?e,
                requirement_id = %requirement.id,
                step_index = idx,
                "Failed to create step"
            );
            e
        })?;
    }

    tracing::info!(
        requirement_id = %requirement.id,
        step_count = parsed_steps.len(),
        "Successfully created requirement with all steps"
    );

    Ok((StatusCode::CREATED, Json(requirement)))
}

pub async fn list_requirements(
    State(pool): State<PgPool>,
    Path(project_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let requirements = requirement_repo::find_requirements_by_project(&pool, project_id).await?;

    // Fetch steps for each requirement
    let mut requirements_with_steps = Vec::new();
    for requirement in requirements {
        let steps = requirement_repo::find_steps_by_requirement(&pool, requirement.id).await?;
        requirements_with_steps.push(RequirementWithSteps { requirement, steps });
    }

    Ok(Json(requirements_with_steps))
}

pub async fn get_requirement(
    State(pool): State<PgPool>,
    Path(requirement_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let requirement = requirement_repo::find_requirement_by_id(&pool, requirement_id)
        .await?
        .ok_or_else(|| AppError::NotFound("Requirement not found".to_string()))?;

    let steps = requirement_repo::find_steps_by_requirement(&pool, requirement_id).await?;

    Ok(Json(RequirementWithSteps { requirement, steps }))
}

pub async fn update_requirement(
    State(pool): State<PgPool>,
    Path(requirement_id): Path<Uuid>,
    Json(payload): Json<UpdateRequirementRequest>,
) -> Result<impl IntoResponse, AppError> {
    payload
        .validate()
        .map_err(|e| AppError::Validation(e.to_string()))?;

    // Validate Gherkin if provided
    if let Some(ref gherkin) = payload.gherkin_scenario {
        GherkinParser::validate(gherkin)
            .map_err(|e| AppError::Validation(format!("Invalid Gherkin: {}", e)))?;
    }

    // Update requirement
    let requirement = requirement_repo::update_requirement(
        &pool,
        requirement_id,
        payload.title.as_deref(),
        payload.description.as_deref(),
        payload.gherkin_scenario.as_deref(),
        payload.position_x,
        payload.position_y,
    )
    .await?;

    // If Gherkin was updated, recreate steps
    if let Some(gherkin) = payload.gherkin_scenario {
        // Delete existing steps
        requirement_repo::delete_steps_by_requirement(&pool, requirement_id).await?;

        // Parse and create new steps
        let parsed_steps = GherkinParser::parse(&gherkin);
        for step in parsed_steps {
            let domain_refs = if step.domain_references.is_empty() {
                None
            } else {
                Some(serde_json::to_value(&step.domain_references).unwrap())
            };

            requirement_repo::create_step(
                &pool,
                requirement_id,
                &step.step_type,
                step.step_order,
                &step.step_text,
                domain_refs.as_ref(),
            )
            .await?;
        }
    }

    Ok(Json(requirement))
}

pub async fn delete_requirement(
    State(pool): State<PgPool>,
    Path(requirement_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    requirement_repo::delete_requirement(&pool, requirement_id).await?;

    Ok(StatusCode::NO_CONTENT)
}

// Connection handlers
pub async fn create_connection(
    State(pool): State<PgPool>,
    Path(project_id): Path<Uuid>,
    Json(payload): Json<CreateConnectionRequest>,
) -> Result<impl IntoResponse, AppError> {
    let source_id = Uuid::parse_str(&payload.source_requirement_id)
        .map_err(|_| AppError::BadRequest("Invalid source requirement ID".to_string()))?;
    let target_id = Uuid::parse_str(&payload.target_requirement_id)
        .map_err(|_| AppError::BadRequest("Invalid target requirement ID".to_string()))?;

    let connection = requirement_repo::create_connection(
        &pool,
        project_id,
        source_id,
        target_id,
        payload.connection_type.as_deref(),
    )
    .await?;

    Ok((StatusCode::CREATED, Json(connection)))
}

pub async fn list_connections(
    State(pool): State<PgPool>,
    Path(project_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    let connections = requirement_repo::find_connections_by_project(&pool, project_id).await?;

    Ok(Json(connections))
}

pub async fn delete_connection(
    State(pool): State<PgPool>,
    Path(connection_id): Path<Uuid>,
) -> Result<impl IntoResponse, AppError> {
    requirement_repo::delete_connection(&pool, connection_id).await?;

    Ok(StatusCode::NO_CONTENT)
}

File: api/routers/requirements.py
"""Requirements router"""
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from pydantic import BaseModel
from typing import List, Optional
import uuid
import re

from api.config.database import get_db
from api.models.requirement import Requirement, RequirementStep, RequirementConnection

router = APIRouter()


class RequirementCreate(BaseModel):
    title: str
    description: Optional[str] = None
    gherkinScenario: str
    positionX: Optional[float] = 0
    positionY: Optional[float] = 0


class RequirementUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    gherkinScenario: Optional[str] = None
    positionX: Optional[float] = None
    positionY: Optional[float] = None


class ConnectionCreate(BaseModel):
    sourceRequirementId: str
    targetRequirementId: str
    connectionType: Optional[str] = "dependency"


def parse_gherkin(gherkin_text: str) -> List[dict]:
    """Parse Gherkin scenario into steps"""
    steps = []
    lines = gherkin_text.strip().split('\n')
    step_order = 0

    for line in lines:
        line = line.strip()
        if not line:
            continue

        # Match Given, When, Then, And, But
        match = re.match(r'^(Given|When|Then|And|But)\s+(.+)$', line, re.IGNORECASE)
        if match:
            step_type = match.group(1).lower()
            step_text = match.group(2)

            # Extract domain references (e.g., {User}, {Order})
            domain_refs = re.findall(r'\{([^}]+)\}', step_text)

            steps.append({
                "stepType": step_type,
                "stepOrder": step_order,
                "stepText": step_text,
                "domainReferences": domain_refs if domain_refs else None
            })
            step_order += 1

    return steps


@router.get("/projects/{project_id}/requirements")
async def list_requirements(project_id: str, db: Session = Depends(get_db)):
    """List all requirements with steps"""
    requirements = db.query(Requirement).filter(Requirement.project_id == uuid.UUID(project_id)).all()

    result = []
    for req in requirements:
        steps = db.query(RequirementStep).filter(
            RequirementStep.requirement_id == req.id
        ).order_by(RequirementStep.step_order).all()

        result.append({
            "requirement": {
                "id": str(req.id),
                "projectId": str(req.project_id),
                "title": req.title,
                "description": req.description,
                "gherkinScenario": req.gherkin_scenario,
                "positionX": req.position_x,
                "positionY": req.position_y,
                "createdAt": req.created_at.isoformat(),
                "updatedAt": req.updated_at.isoformat()
            },
            "steps": [
                {
                    "id": str(step.id),
                    "requirementId": str(step.requirement_id),
                    "stepType": step.step_type,
                    "stepOrder": step.step_order,
                    "stepText": step.step_text,
                    "domainReferences": step.domain_references,
                    "createdAt": step.created_at.isoformat()
                }
                for step in steps
            ]
        })

    return result


@router.post("/projects/{project_id}/requirements", status_code=201)
async def create_requirement(
    project_id: str,
    requirement: RequirementCreate,
    db: Session = Depends(get_db)
):
    """Create new requirement with Gherkin parsing"""
    # Create requirement
    new_req = Requirement(
        project_id=uuid.UUID(project_id),
        title=requirement.title,
        description=requirement.description,
        gherkin_scenario=requirement.gherkinScenario,
        position_x=requirement.positionX,
        position_y=requirement.positionY
    )
    db.add(new_req)
    db.flush()  # Get the ID without committing

    # Parse Gherkin and create steps
    parsed_steps = parse_gherkin(requirement.gherkinScenario)
    steps = []

    for step_data in parsed_steps:
        step = RequirementStep(
            requirement_id=new_req.id,
            step_type=step_data["stepType"],
            step_order=step_data["stepOrder"],
            step_text=step_data["stepText"],
            domain_references=step_data["domainReferences"]
        )
        db.add(step)
        steps.append(step)

    db.commit()
    db.refresh(new_req)

    # Refresh all steps
    for step in steps:
        db.refresh(step)

    return {
        "requirement": {
            "id": str(new_req.id),
            "projectId": str(new_req.project_id),
            "title": new_req.title,
            "description": new_req.description,
            "gherkinScenario": new_req.gherkin_scenario,
            "positionX": new_req.position_x,
            "positionY": new_req.position_y,
            "createdAt": new_req.created_at.isoformat(),
            "updatedAt": new_req.updated_at.isoformat()
        },
        "steps": [
            {
                "id": str(step.id),
                "requirementId": str(step.requirement_id),
                "stepType": step.step_type,
                "stepOrder": step.step_order,
                "stepText": step.step_text,
                "domainReferences": step.domain_references,
                "createdAt": step.created_at.isoformat()
            }
            for step in steps
        ]
    }


@router.get("/requirements/{requirement_id}")
async def get_requirement(requirement_id: str, db: Session = Depends(get_db)):
    """Get requirement with steps"""
    req = db.query(Requirement).filter(Requirement.id == uuid.UUID(requirement_id)).first()
    if not req:
        raise HTTPException(status_code=404, detail="Requirement not found")

    steps = db.query(RequirementStep).filter(
        RequirementStep.requirement_id == req.id
    ).order_by(RequirementStep.step_order).all()

    return {
        "requirement": {
            "id": str(req.id),
            "projectId": str(req.project_id),
            "title": req.title,
            "description": req.description,
            "gherkinScenario": req.gherkin_scenario,
            "positionX": req.position_x,
            "positionY": req.position_y,
            "createdAt": req.created_at.isoformat(),
            "updatedAt": req.updated_at.isoformat()
        },
        "steps": [
            {
                "id": str(step.id),
                "requirementId": str(step.requirement_id),
                "stepType": step.step_type,
                "stepOrder": step.step_order,
                "stepText": step.step_text,
                "domainReferences": step.domain_references,
                "createdAt": step.created_at.isoformat()
            }
            for step in steps
        ]
    }


@router.put("/requirements/{requirement_id}")
async def update_requirement(
    requirement_id: str,
    requirement: RequirementUpdate,
    db: Session = Depends(get_db)
):
    """Update requirement"""
    req = db.query(Requirement).filter(Requirement.id == uuid.UUID(requirement_id)).first()
    if not req:
        raise HTTPException(status_code=404, detail="Requirement not found")

    # Update fields
    if requirement.title is not None:
        req.title = requirement.title
    if requirement.description is not None:
        req.description = requirement.description
    if requirement.positionX is not None:
        req.position_x = requirement.positionX
    if requirement.positionY is not None:
        req.position_y = requirement.positionY

    # If Gherkin changed, reparse and update steps
    if requirement.gherkinScenario is not None:
        req.gherkin_scenario = requirement.gherkinScenario

        # Delete existing steps
        db.query(RequirementStep).filter(RequirementStep.requirement_id == req.id).delete()

        # Parse and create new steps
        parsed_steps = parse_gherkin(requirement.gherkinScenario)
        for step_data in parsed_steps:
            step = RequirementStep(
                requirement_id=req.id,
                step_type=step_data["stepType"],
                step_order=step_data["stepOrder"],
                step_text=step_data["stepText"],
                domain_references=step_data["domainReferences"]
            )
            db.add(step)

    db.commit()
    db.refresh(req)

    # Get updated steps
    steps = db.query(RequirementStep).filter(
        RequirementStep.requirement_id == req.id
    ).order_by(RequirementStep.step_order).all()

    return {
        "requirement": {
            "id": str(req.id),
            "projectId": str(req.project_id),
            "title": req.title,
            "description": req.description,
            "gherkinScenario": req.gherkin_scenario,
            "positionX": req.position_x,
            "positionY": req.position_y,
            "createdAt": req.created_at.isoformat(),
            "updatedAt": req.updated_at.isoformat()
        },
        "steps": [
            {
                "id": str(step.id),
                "requirementId": str(step.requirement_id),
                "stepType": step.step_type,
                "stepOrder": step.step_order,
                "stepText": step.step_text,
                "domainReferences": step.domain_references,
                "createdAt": step.created_at.isoformat()
            }
            for step in steps
        ]
    }


@router.delete("/requirements/{requirement_id}", status_code=204)
async def delete_requirement(requirement_id: str, db: Session = Depends(get_db)):
    """Delete requirement"""
    req = db.query(Requirement).filter(Requirement.id == uuid.UUID(requirement_id)).first()
    if not req:
        raise HTTPException(status_code=404, detail="Requirement not found")

    db.delete(req)
    db.commit()
    return None


# Requirement Connections endpoints
@router.get("/projects/{project_id}/connections")
async def list_connections(project_id: str, db: Session = Depends(get_db)):
    """List all requirement connections"""
    connections = db.query(RequirementConnection).filter(
        RequirementConnection.project_id == uuid.UUID(project_id)
    ).all()

    return [
        {
            "id": str(conn.id),
            "projectId": str(conn.project_id),
            "sourceRequirementId": str(conn.source_requirement_id),
            "targetRequirementId": str(conn.target_requirement_id),
            "connectionType": conn.connection_type,
            "createdAt": conn.created_at.isoformat()
        }
        for conn in connections
    ]


@router.post("/projects/{project_id}/connections", status_code=201)
async def create_connection(
    project_id: str,
    connection: ConnectionCreate,
    db: Session = Depends(get_db)
):
    """Create requirement connection"""
    new_conn = RequirementConnection(
        project_id=uuid.UUID(project_id),
        source_requirement_id=uuid.UUID(connection.sourceRequirementId),
        target_requirement_id=uuid.UUID(connection.targetRequirementId),
        connection_type=connection.connectionType
    )
    db.add(new_conn)
    db.commit()
    db.refresh(new_conn)

    return {
        "id": str(new_conn.id),
        "projectId": str(new_conn.project_id),
        "sourceRequirementId": str(new_conn.source_requirement_id),
        "targetRequirementId": str(new_conn.target_requirement_id),
        "connectionType": new_conn.connection_type,
        "createdAt": new_conn.created_at.isoformat()
    }


@router.delete("/connections/{connection_id}", status_code=204)
async def delete_connection(connection_id: str, db: Session = Depends(get_db)):
    """Delete connection"""
    conn = db.query(RequirementConnection).filter(
        RequirementConnection.id == uuid.UUID(connection_id)
    ).first()
    if not conn:
        raise HTTPException(status_code=404, detail="Connection not found")

    db.delete(conn)
    db.commit()
    return None
